### PR TITLE
Gas Lift Test Cases

### PR DESCRIPTION
**Pull request by OPMUSER**
WIP Gas Lift Test Cases, see GASLIFT.md for case descriptions. More cases to follow and additional comparisons.

### CODE REVIEW

**Comment by alfbr** [comment]
Thank you, these are highly appreciated! I have not run them yet, but they seem well suited. This is ready for merging, but please consider making the progressive complexity of the tests clear in the DATA-file headers. That is, in the second test case you can explicitly state that the case differs from the first by introducing irregular cells. Similarly for the rest, a short sentence can be added explaining how it differs from the preceding case. While it can be guessed from the md-file, it is very hard to see it from the file header. I will add lines explaining how this can look.

File: gaslift/GASLIFT-02.DATA [hunk 1]
```diff
@@ -0,0 +1,1154 @@
+-- This reservoir simulation deck is made available under the Open Database
+-- License: http://opendatacommons.org/licenses/odbl/1.0/. Any rights in
+-- individual contents of the database are licensed under the Database Contents
+-- License: http://opendatacommons.org/licenses/dbcl/1.0/
+  
+-- Copyright (C) 2021 Equinor
+
```
Comment (alfbr) [inline gaslift/GASLIFT-02.DATA:7]: -- This model is based on GASLIFT-01.DATA, it differs by introducing an irregular corner-point geometry.

File: gaslift/GASLIFT-03.DATA [hunk 2]
```diff
@@ -0,0 +1,1116 @@
+-- This reservoir simulation deck is made available under the Open Database
+-- License: http://opendatacommons.org/licenses/odbl/1.0/. Any rights in
+-- individual contents of the database are licensed under the Database Contents
+-- License: http://opendatacommons.org/licenses/dbcl/1.0/
+  
+-- Copyright (C) 2021 Equinor
+
```
Comment (alfbr) [inline gaslift/GASLIFT-03.DATA:7]: -- This model is based on GASLIFT-02.DATA, it differs by introducing a three phase model with more complex PVT.

File: gaslift/GASLIFT-04.DATA [hunk 3]
```diff
@@ -0,0 +1,959 @@
+-- This reservoir simulation deck is made available under the Open Database
+-- License: http://opendatacommons.org/licenses/odbl/1.0/. Any rights in
+-- individual contents of the database are licensed under the Database Contents
+-- License: http://opendatacommons.org/licenses/dbcl/1.0/
+  
+-- Copyright (C) 2021 Equinor
+
```
Comment (alfbr) [inline gaslift/GASLIFT-04.DATA:7]: -- This model is based on GASLIFT-03.DATA, it differs by introducing lift curves in VFP tables.

File: gaslift/GASLIFT-03.DATA [hunk 4]
```diff
@@ -0,0 +1,1116 @@
...
+UNIFIN                                                                          
+--
+--       SWITCH ON THE UNIFIED OUTPUT FILES OPTION
+--
+UNIFOUT                                                                          
+--
+--       PATH       PATH
+--       ALIAS      DIRECTORY FILENAME
+PATHS
+        'MODEL05'   '..\model5\include'                                        /
```
Comment (alfbr) [inline gaslift/GASLIFT-03.DATA:150]: Please create a new folder named include here, and copy over the include files from the model5 folder.

**Comment by alfbr** [comment]
Just one more thing, there are include paths to the neighboring model5 folder. I suggest all folders in opm-tests are self-contained, so please create a folder named include under the gaslift folder, and copy all include files you need into it.

File: gaslift/GASLIFT-03.DATA (hunk 4 above)
Comment (OPMUSER) [reply gaslift/GASLIFT-03.DATA:152]: I'll like to keep this for the time being, as I may use different models in generating the tests. We already have a couple of gas lifts in model5 and I was considering moving these any any others gas lifts tests under the same directory.

File: gaslift/GASLIFT-03.DATA (hunk 2 above)
Comment (OPMUSER) [reply gaslift/GASLIFT-03.DATA:7]: Done

File: gaslift/GASLIFT-02.DATA (hunk 1 above)
Comment (OPMUSER) [reply gaslift/GASLIFT-02.DATA:7]: Done

File: gaslift/GASLIFT-03.DATA (hunk 4 above)
Comment (alfbr) [reply gaslift/GASLIFT-03.DATA:152]: Then I believe you should move them over in a not to soon future. People working in one folder may change include files at their will without knowing that they are in use in other places in the repo. I am open to being wrong about this, but then we should have a broader discussion where the other active contributors voice in. My preference would be that you copy the include files over now, and then just delete them again if you move the tests over to the model5 folder. In any case, I have made my point, and will respect your wish at this point.

**Comment by alfbr** [comment]
Please also update 5, 6 and 7 with a short sentence explaining the difference from the preceding case, and this is ready to be merged.

**Comment by alfbr** [comment]
Please give me a ping when you are happy with the updates. From my end this is ready for merging now.

**Comment by OPMUSER** [comment]
Noted.

I want to add a few more tests in the next few days and will let you know when ready.

**Comment by OPMUSER** [comment]
Ready for a review.

Note:

- There are 15 test cases that test various aspects of gas lift optimization, unfortunately none of the cases match E100. There are two gas lift cases in MODEL5, that I have not exactly duplicated, I could if required, or we could just delete these cases and use a few of the new 15 cases for regression testing. Comments appreciated.
- There are lot of issues with gas lift optimization; however, I think some of these are caused by well convergence issues and how we handle this, it appears after a well convergence issue that gas lift stops working for the offending well. There may be other issues as well.
- Sometimes, as per GASLIFT-05 the field results appear reasonable for most of the run; however, when we look at the individual wells flowing the results are different. Perhaps OPM Flow's gas lift calculations are not close enough to re-produce the results of E100, or perhaps how we handle well convergence is causing the issue. E100 has extensive debug output on both the well solution and gas lift optimization, and may be of use, although understanding what the numbers are will require some effort.
- The GASLIFT-10 run is interesting because all the wells die but WTEST brings on the wells again and they seem to behave as expected after that. Perharps that is a clue of what is going on.
-  Running with the Extended Network model is problematic.
- There has been some discussions with @atgeirr and @joakim-hove concerning the opm-test directory structure, and there is some agreement that my suggested organization will be beneficial in the long term. Just to recap, what I propose is that we have just model and topic directories. In the model direct we have the base model, no other tests just the base run. And in the topic directories we "include" the files we need from the model directories. This is a cleaner implementation and we avoid duplication.I'm not proposing that we do this straight away but that we move to this approach over time. The advantage, is that topic directories can use different models to test the given feature. For example we could now delete the gas lift tests in the model5 directory (or duplicate them in the GASLIFT topic directory.

Feed back appreciated.

**Comment by OPMUSER** [comment]
PS: I forgot to mention that all the gas lift cases include in the model5 files using:

```
--
--       PATH       PATH
--       ALIAS      DIRECTORY FILENAME
PATHS
        'MODEL05'   '../model5/include'                                        /
/

```
and then using

```
--
--       LOAD INCLUDE FILE
--
INCLUDE 
         '$MODEL05/test1_20x30x10.grdecl'         /

```
to load the required file

**Comment by bska** [comment]
jenkins build this please

**Comment by alfbr** [comment]
Thank you, merging this now. Please also consider shortening down the run time of these data sets prior to including them as regression tests.

**Pull request merged**